File: Packages/TeamTasksKit/Sources/TeamTasksCore/Logic/InputValidation.swift
import Foundation

/// Input rules of docs/CONTRACTS.md §1 and docs/CONTRACTS-V2.md §1, §3, §5, shared by the mock backend and the
/// Supabase adapters.
///
/// Adapters apply them before calling the server, so both backends answer the same even where the server alone
/// would not: Supabase Auth accepts a blank display name (the profile trigger falls back to the e-mail), does not
/// trim e-mails, and Postgres rejects U+0000 with a generic `22P05`.
///
/// v2 order of `create_task` (docs/CONTRACTS-V2.md §3): title → details → due date → `recurrence(_:dueAt:)` →
/// `rotation(_:recurrence:)` then its membership → assignees (without rotation: count, then membership) →
/// `checklist(_:)`. The membership rules need the group's members: an adapter that leaves them to the server must not
/// check anything that comes after them (the checklist) on the client, or its first error could differ.
public enum InputValidation {
    /// Code points trimmed at both ends of every text field: U+0009–U+000D, U+0020, U+0085, U+00A0, U+1680,
    /// U+2000–U+200B, U+2028, U+2029, U+202F, U+205F, U+3000.
    ///
    /// An explicit list because `CharacterSet.whitespacesAndNewlines` differs between Linux and Apple platforms
    /// (U+200B). Must match the class of the SQL `private.clean_text`.
    public static let trimmedCharacters: CharacterSet = {
        let ranges: [ClosedRange<Unicode.Scalar>] = [
            "\u{09}"..."\u{0D}", "\u{20}"..."\u{20}", "\u{85}"..."\u{85}", "\u{A0}"..."\u{A0}",
            "\u{1680}"..."\u{1680}", "\u{2000}"..."\u{200B}", "\u{2028}"..."\u{2029}", "\u{202F}"..."\u{202F}",
            "\u{205F}"..."\u{205F}", "\u{3000}"..."\u{3000}",
        ]
        var set = CharacterSet()
        for range in ranges {
            set.insert(charactersIn: range)
        }
        return set
    }()

    /// Supabase Auth hashes passwords with bcrypt, which reads at most 72 bytes.
    public static let passwordMaxBytes = 72
    /// Supabase Auth refuses longer e-mail addresses.
    public static let emailMaxBytes = 255

    // MARK: - Text fields

    /// `value` without the `trimmedCharacters` at both ends (code point by code point, like the SQL regexp).
    public static func trimmed(_ value: String) -> String {
        let scalars = value.unicodeScalars
        guard let first = scalars.firstIndex(where: { !trimmedCharacters.contains($0) }),
              let last = scalars.lastIndex(where: { !trimmedCharacters.contains($0) })
        else { return "" }
        return String(scalars[first...last])
    }

    /// Length as Postgres `char_length` counts it: Unicode scalars.
    public static func length(_ value: String) -> Int {
        value.unicodeScalars.count
    }

    /// Trimmed display name (1–50) or `.invalidDisplayName`.
    public static func displayName(_ raw: String) throws -> String {
        try text(raw, lengths: Limits.displayName, error: .invalidDisplayName)
    }

    /// Trimmed group name (1–60) or `.invalidName`.
    public static func groupName(_ raw: String) throws -> String {
        try text(raw, lengths: Limits.groupName, error: .invalidName)
    }

    /// Trimmed task title (1–200) or `.invalidTitle`.
    public static func taskTitle(_ raw: String) throws -> String {
        try text(raw, lengths: Limits.taskTitle, error: .invalidTitle)
    }

    /// Trimmed task details (≤ 5000, nil when empty) or `.invalidDetails`.
    public static func taskDetails(_ raw: String) throws -> String? {
        let value = try text(raw, lengths: 0...Limits.taskDetailsMax, error: .invalidDetails)
        return value.isEmpty ? nil : value
    }

    /// Accepted due dates (`tasks_due_at_range`): [1970-01-01, 10000-01-01) UTC. Excludes ±infinity and BC
    /// dates, which PostgREST renders in forms no JSON decoder accepts.
    public static let dueDateRange: Range<Date> =
        Date(timeIntervalSince1970: 0)..<Date(timeIntervalSince1970: 253_402_300_800)

    /// nil (no due date) or a date inside `dueDateRange`, else `.invalidInput` (SQL `invalid_due_at`).
    /// Checked after the title and the details, before the assignees (same order as the SQL).
    public static func dueDate(_ date: Date?) throws -> Date? {
        guard let date else { return nil }
        guard dueDateRange.contains(date) else { throw AppError.invalidInput }
        return date
    }

    /// Postgres `text` cannot hold U+0000: such input gets the field's validation error.
    private static func text(_ raw: String, lengths: ClosedRange<Int>, error: AppError) throws -> String {
        let value = trimmed(raw)
        guard lengths.contains(length(value)), !value.unicodeScalars.contains("\u{0}") else { throw error }
        return value
    }

    // MARK: - Auth

    /// Trimmed and lowercased e-mail (Supabase Auth stores e-mails lowercased but does not trim them, so the
    /// app normalizes every e-mail it sends to Auth).
    public static func normalizedEmail(_ raw: String) -> String {
        trimmed(raw).lowercased()
    }

    /// The normalized e-mail, or `.invalidEmail` unless it has the HTML5 syntax Supabase Auth checks: before the
    /// `@`, one or more ASCII letters, digits or characters among .!#$%&'*+/=?^_{|}~- and the backtick; after
    /// it, dot-separated labels of 1–63 ASCII letters, digits or hyphens, starting and ending with a letter or
    /// digit (`user@localhost` is valid). At most 255 bytes.
    public static func email(_ raw: String) throws -> String {
        let value = trimmed(raw)
        guard value.utf8.count <= emailMaxBytes, hasEmailSyntax(value) else { throw AppError.invalidEmail }
        return value.lowercased()
    }

    /// Password length in UTF-8 bytes, like Supabase Auth: at least `Limits.passwordMinLength` (`.weakPassword`),
    /// at most `passwordMaxBytes` (`.invalidInput`).
    public static func password(_ value: String) throws {
        let bytes = value.utf8.count
        guard bytes >= Limits.passwordMinLength else { throw AppError.weakPassword }
        guard bytes <= passwordMaxBytes else { throw AppError.invalidInput }
    }

    /// Sign-up input, checked in this order: e-mail, password, display name. Returns the normalized e-mail and
    /// the trimmed display name.
    public static func signUp(email: String, password: String, displayName: String) throws -> (email: String, displayName: String) {
        let email = try self.email(email)
        try self.password(password)
        let name = try self.displayName(displayName)
        return (email, name)
    }

    private static let emailLocalCharacters = Set("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789.!#$%&'*+/=?^_`{|}~-".unicodeScalars)

    private static func hasEmailSyntax(_ value: String) -> Bool {
        let scalars = Array(value.unicodeScalars)
        guard let at = scalars.firstIndex(of: "@"), at > 0 else { return false }
        guard scalars[..<at].allSatisfy({ emailLocalCharacters.contains($0) }) else { return false }
        let labels = scalars[(at + 1)...].split(separator: ".", omittingEmptySubsequences: false)
        return labels.allSatisfy { label in
            guard (1...63).contains(label.count),
                  let first = label.first, let last = label.last,
                  isASCIIAlphanumeric(first), isASCIIAlphanumeric(last)
            else { return false }
            return label.allSatisfy { isASCIIAlphanumeric($0) || $0 == "-" }
        }
    }

    private static func isASCIIAlphanumeric(_ scalar: Unicode.Scalar) -> Bool {
        ("a"..."z").contains(scalar) || ("A"..."Z").contains(scalar) || ("0"..."9").contains(scalar)
    }

    // MARK: - v2: appearance (docs/CONTRACTS-V2.md §1)

    /// A color key from its stored text, exact case (`coral`); nil = automatic. `Coral`, `""` or an unknown key throw
    /// `.invalidAppearance` (SQL `invalid_color`). Typed `ColorKey` values are always valid.
    public static func colorKey(_ raw: String?) throws -> ColorKey? {
        guard let raw else { return nil }
        guard let key = ColorKey(rawValue: raw) else { throw AppError.invalidAppearance }
        return key
    }

    /// The normalized emoji of a group or an avatar (SQL `private.normalize_emoji`): nil stays nil; otherwise
    /// `trimmed(_:)`, and an empty result is nil. What remains must have 1–`Limits.emojiCodePointsMax` code points,
    /// none of them a trimmed code point or a C0/C1 control (`isForbiddenInEmoji`), else `.invalidAppearance`
    /// (SQL `invalid_emoji`). ZWJ sequences, variation selectors, keycaps and flags are accepted.
    public static func emoji(_ raw: String?) throws -> String? {
        guard let raw else { return nil }
        let value = trimmed(raw)
        guard !value.isEmpty else { return nil }
        guard value.unicodeScalars.count <= Limits.emojiCodePointsMax,
              !value.unicodeScalars.contains(where: isForbiddenInEmoji)
        else { throw AppError.invalidAppearance }
        return value
    }

    /// Code points refused inside an emoji: U+0000–U+0020, U+007F–U+00A0, U+1680, U+2000–U+200B, U+2028, U+2029,
    /// U+202F, U+205F, U+3000 (the trimmed code points and the C0/C1 controls; U+200D, the ZWJ, is allowed).
    public static func isForbiddenInEmoji(_ scalar: Unicode.Scalar) -> Bool {
        scalar.value <= 0x20 || (0x7F...0xA0).contains(scalar.value) || trimmedCharacters.contains(scalar)
    }

    // MARK: - v2: recurrence (docs/CONTRACTS-V2.md §5)

    /// Zone names the server accepts that Foundation may list or name differently (`UTC` can read as `GMT`).
    private static let fixedTimeZoneIds: Set<String> = ["UTC", "GMT", "Etc/UTC", "Etc/GMT"]
    private static let knownTimeZoneIds = Set(TimeZone.knownTimeZoneIdentifiers)

    /// True for a time zone name accepted in a recurrence rule: an IANA name that this device knows, exact case
    /// (`Europe/Paris`, `UTC`, `America/Argentina/Buenos_Aires`), except the `posix/…` and `right/…` copies and
    /// `Factory`. POSIX offsets (`UTC+3`), abbreviations (`CEST`) and other cases (`europe/paris`) are refused.
    /// The server checks the same names against PostgreSQL's `pg_timezone_names`.
    public static func isValidTimeZoneId(_ id: String) -> Bool {
        guard !id.hasPrefix("posix/"), !id.hasPrefix("right/"), id != "Factory" else { return false }
        guard knownTimeZoneIds.contains(id) || fixedTimeZoneIds.contains(id) else { return false }
        return TimeZone(identifier: id) != nil
    }

    /// The shape of a rule, else `.invalidRecurrence` (SQL `invalid_recurrence`): `interval` in
    /// 1…`Limits.repeatIntervalMax`; `weekdays` only on a weekly rule, not empty, values in 1…7; `timeZoneId`
    /// accepted by `isValidTimeZoneId(_:)`. `monthDay` is not checked: the server sets it.
    public static func recurrence(_ rule: RecurrenceRule) throws -> RecurrenceRule {
        guard (1...Limits.repeatIntervalMax).contains(rule.interval) else { throw AppError.invalidRecurrence }
        if let weekdays = rule.weekdays {
            guard rule.frequency == .weekly, !weekdays.isEmpty, weekdays.allSatisfy({ (1...7).contains($0) }) else {
                throw AppError.invalidRecurrence
            }
        }
        guard isValidTimeZoneId(rule.timeZoneId) else { throw AppError.invalidRecurrence }
        return rule
    }

    /// The recurrence of a draft: nil (no repetition), or the rule's shape (`recurrence(_:)`), then a due date is
    /// required (`.recurrenceNeedsDueDate`). Checked after the due date, before the rotation.
    public static func recurrence(_ rule: RecurrenceRule?, dueAt: Date?) throws -> RecurrenceRule? {
        guard let rule else { return nil }
        let checked = try recurrence(rule)
        guard dueAt != nil else { throw AppError.recurrenceNeedsDueDate }
        return checked
    }

    /// A new rotation, except the membership of its users (checked by the backend, with the same error): empty is no
    /// rotation; otherwise it needs a `recurrence`, and `Limits.rotationMin`…`Limits.rotationMax` distinct ids, else
    /// `.invalidRotation`. Checked after the recurrence, before the assignees (which a rotation replaces).
    /// On update, a rotation equal to the task's current one is kept without any check.
    public static func rotation(_ ids: [UUID], recurrence: RecurrenceRule?) throws -> [UUID] {
        guard !ids.isEmpty else { return [] }
        guard recurrence != nil,
              (Limits.rotationMin...Limits.rotationMax).contains(ids.count),
              Set(ids).count == ids.count
        else { throw AppError.invalidRotation }
        return ids
    }

    // MARK: - v2: checklist (docs/CONTRACTS-V2.md §3)

    /// Trimmed checklist item title (1–`Limits.checklistItemTitleMax`), else `.invalidChecklistItem`
    /// (SQL `invalid_item_title`).
    public static func checklistItemTitle(_ raw: String) throws -> String {
        try text(raw, lengths: 1...Limits.checklistItemTitleMax, error: .invalidChecklistItem)
    }

    /// The initial checklist of a new task: every title in order (`checklistItemTitle(_:)`), then at most
    /// `Limits.checklistItemsMax` items (`.tooManyChecklistItems`). Checked last, after the assignees.
    public static func checklist(_ titles: [String]) throws -> [String] {
        let checked = try titles.map { try checklistItemTitle($0) }
        guard checked.count <= Limits.checklistItemsMax else { throw AppError.tooManyChecklistItems }
        return checked
    }
}

File: android/core/src/main/kotlin/io/github/notkanaa/equipe/core/InputValidation.kt
package io.github.notkanaa.equipe.core

import java.time.Instant

/**
 * Input rules of docs/CONTRACTS.md §1, shared by the mock backend and the Supabase adapters (port of
 * TeamTasksCore/Logic/InputValidation.swift).
 *
 * Adapters apply them before calling the server, so both backends answer the same even where the server alone
 * would not: Supabase Auth accepts a blank display name (the profile trigger falls back to the e-mail), does not
 * trim e-mails, and Postgres rejects U+0000 with a generic `22P05`.
 *
 * Every function works on code points (Unicode scalars), like Postgres and Swift, never on UTF-16 units.
 */
object InputValidation {
    /**
     * Code points trimmed at both ends of every text field: U+0009–U+000D, U+0020, U+0085, U+00A0, U+1680,
     * U+2000–U+200B, U+2028, U+2029, U+202F, U+205F, U+3000.
     *
     * An explicit list because every platform's "whitespace" differs (U+200B, U+0085, U+001C–U+001F…).
     * Must match the class of the SQL `private.clean_text`.
     */
    val trimmedCodePoints: Set<Int> = buildSet {
        addAll(0x09..0x0D)
        add(0x20)
        add(0x85)
        add(0xA0)
        add(0x1680)
        addAll(0x2000..0x200B)
        add(0x2028)
        add(0x2029)
        add(0x202F)
        add(0x205F)
        add(0x3000)
    }

    /** Supabase Auth hashes passwords with bcrypt, which reads at most 72 bytes. */
    const val passwordMaxBytes: Int = 72

    /** Supabase Auth refuses longer e-mail addresses. */
    const val emailMaxBytes: Int = 255

    // region Text fields

    /** [value] without the [trimmedCodePoints] at both ends (code point by code point, like the SQL regexp). */
    fun trimmed(value: String): String {
        var start = 0
        while (start < value.length) {
            val codePoint = value.codePointAt(start)
            if (codePoint !in trimmedCodePoints) break
            start += Character.charCount(codePoint)
        }
        var end = value.length
        while (end > start) {
            val codePoint = value.codePointBefore(end)
            if (codePoint !in trimmedCodePoints) break
            end -= Character.charCount(codePoint)
        }
        return value.substring(start, end)
    }

    /** Length as Postgres `char_length` counts it: code points. */
    fun length(value: String): Int = value.codePointLength()

    /** Trimmed display name (1–50) or [AppError.InvalidDisplayName]. */
    fun displayName(raw: String): String = text(raw, Limits.displayName, AppError.InvalidDisplayName)

    /** Trimmed group name (1–60) or [AppError.InvalidName]. */
    fun groupName(raw: String): String = text(raw, Limits.groupName, AppError.InvalidName)

    /** Trimmed task title (1–200) or [AppError.InvalidTitle]. */
    fun taskTitle(raw: String): String = text(raw, Limits.taskTitle, AppError.InvalidTitle)

    /** Trimmed task details (≤ 5000, null when empty) or [AppError.InvalidDetails]. */
    fun taskDetails(raw: String): String? {
        val value = text(raw, 0..Limits.taskDetailsMax, AppError.InvalidDetails)
        return value.ifEmpty { null }
    }

    /**
     * Accepted due dates (`tasks_due_at_range`): [1970-01-01, 10000-01-01) UTC. Excludes ±infinity and BC dates,
     * which PostgREST renders in forms no JSON decoder accepts.
     */
    val dueDateRange: OpenEndRange<Instant> = Instant.EPOCH..<Instant.ofEpochSecond(253_402_300_800L)

    /**
     * null (no due date) or a date inside [dueDateRange], else [AppError.InvalidInput] (SQL `invalid_due_at`).
     * Checked after the title and the details, before the assignees (same order as the SQL).
     */
    fun dueDate(date: Instant?): Instant? {
        if (date == null) return null
        if (date !in dueDateRange) throw AppError.InvalidInput
        return date
    }

    /** Postgres `text` cannot hold U+0000: such input gets the field's validation error. */
    private fun text(raw: String, lengths: IntRange, error: AppError): String {
        val value = trimmed(raw)
        if (length(value) !in lengths || value.indexOf('\u0000') >= 0) throw error
        return value
    }

    // endregion

    // region Auth

    /**
     * Trimmed and lowercased e-mail (Supabase Auth stores e-mails lowercased but does not trim them, so the app
     * normalizes every e-mail it sends to Auth).
     */
    fun normalizedEmail(raw: String): String = trimmed(raw).lowercase()

    /**
     * The normalized e-mail, or [AppError.InvalidEmail] unless it has the HTML5 syntax Supabase Auth checks: before
     * the `@`, one or more ASCII letters, digits or characters among .!#$%&'*+/=?^_{|}~- and the backtick; after it,
     * dot-separated labels of 1–63 ASCII letters, digits or hyphens, starting and ending with a letter or digit
     * (`user@localhost` is valid). At most 255 bytes.
     */
    fun email(raw: String): String {
        val value = trimmed(raw)
        if (utf8Length(value) > emailMaxBytes || !hasEmailSyntax(value)) throw AppError.InvalidEmail
        return value.lowercase()
    }

    /**
     * Password length in UTF-8 bytes, like Supabase Auth: at least [Limits.passwordMinLength]
     * ([AppError.WeakPassword]), at most [passwordMaxBytes] ([AppError.InvalidInput]).
     */
    fun password(value: String) {
        val bytes = utf8Length(value)
        if (bytes < Limits.passwordMinLength) throw AppError.WeakPassword
        if (bytes > passwordMaxBytes) throw AppError.InvalidInput
    }

    /** Normalized sign-up input (Swift returns the tuple `(email:, displayName:)`). */
    data class SignUpInput(val email: String, val displayName: String)

    /**
     * Sign-up input, checked in this order: e-mail, password, display name. Returns the normalized e-mail and the
     * trimmed display name.
     */
    fun signUp(email: String, password: String, displayName: String): SignUpInput {
        val normalizedEmail = email(email)
        password(password)
        val name = displayName(displayName)
        return SignUpInput(normalizedEmail, name)
    }

    private val emailLocalCodePoints: Set<Int> =
        "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789.!#$%&'*+/=?^_`{|}~-".map { it.code }.toSet()

    private fun hasEmailSyntax(value: String): Boolean {
        val codePoints = value.codePoints().toArray()
        val at = codePoints.indexOf('@'.code)
        if (at <= 0) return false
        for (index in 0 until at) {
            if (codePoints[index] !in emailLocalCodePoints) return false
        }
        // Split the domain on "." keeping empty labels (Swift `omittingEmptySubsequences: false`).
        val labels = mutableListOf<List<Int>>()
        var current = mutableListOf<Int>()
        for (index in at + 1 until codePoints.size) {
            val codePoint = codePoints[index]
            if (codePoint == '.'.code) {
                labels.add(current)
                current = mutableListOf()
            } else {
                current.add(codePoint)
            }
        }
        labels.add(current)
        return labels.all { label ->
            label.size in 1..63 &&
                isAsciiAlphanumeric(label.first()) &&
                isAsciiAlphanumeric(label.last()) &&
                label.all { isAsciiAlphanumeric(it) || it == '-'.code }
        }
    }

    private fun isAsciiAlphanumeric(codePoint: Int): Boolean =
        codePoint in 'a'.code..'z'.code || codePoint in 'A'.code..'Z'.code || codePoint in '0'.code..'9'.code

    private fun utf8Length(value: String): Int = value.toByteArray(Charsets.UTF_8).size

    // endregion
}
